locust/test/test_main.py

```diff
@@ -2213,25 +2631,52 @@ def my_task(self):
                     "--processes",
                     "2",
                     "--headless",
+                    "--master-port",
+                    str(master_port),
                 ],
                 stdout=PIPE,
                 stderr=PIPE,
                 text=True,
+                bufsize=1,
                 start_new_session=True,
             )
-            gevent.sleep(2)
-            master_proc.kill()
-            master_proc.wait()
+
             try:
-                worker_stdout, worker_stderr = worker_parent_proc.communicate(timeout=7)
-            except Exception:
-                os.killpg(worker_parent_proc.pid, signal.SIGTERM)
+
+                def master_ready():
+                    line = master_proc.stderr.readline()
+                    return "All users spawned" in line
+
+                poll_until(master_ready, timeout=60, sleep_time=0.1)
+
+                master_proc.terminate()
+                try:
+                    master_proc.wait(timeout=5)
+                except subprocess.TimeoutExpired:
+                    master_proc.kill()
+                    master_proc.wait()
+
+                def workers_shut_down():
+                    return worker_parent_proc.poll() is not None
+
+                poll_until(workers_shut_down, timeout=60, sleep_time=0.1)
+
                 worker_stdout, worker_stderr = worker_parent_proc.communicate()
-                assert False, f"worker never finished: {worker_stdout} / {worker_stderr}"
 
-            self.assertNotIn("Traceback", worker_stderr)
-            self.assertIn("Didn't get heartbeat from master in over ", worker_stderr)
-            self.assertIn("worker index:", worker_stdout)
+                self.assertNotIn("Traceback", worker_stderr)
+                self.assertIn("Got quit message from master, shutting down", worker_stderr)
```

Comment:
Oh. I'm afraid you broke this test. It is for validating that if master is _killed_ (not politely shut down using `.terminate()`) , the workers will shut down.